// react-native-css-interop/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rem = exports.vars = exports.useUnstableNativeVariable = exports.useColorScheme = exports.colorScheme = exports.StyleSheet = exports.createInteropElement = exports.remapProps = exports.cssInterop = exports.wrapJSX = exports.createElement = void 0;
var react_1 = require("react");
Object.defineProperty(exports, "createElement", { enumerable: true, get: function () { return react_1.createElement; } });
var wrap_jsx_1 = require("./runtime/wrap-jsx");
Object.defineProperty(exports, "wrapJSX", { enumerable: true, get: function () { return __importDefault(wrap_jsx_1).default; } });
var runtime_1 = require("./runtime");
Object.defineProperty(exports, "cssInterop", { enumerable: true, get: function () { return runtime_1.cssInterop; } });
Object.defineProperty(exports, "remapProps", { enumerable: true, get: function () { return runtime_1.remapProps; } });
Object.defineProperty(exports, "createInteropElement", { enumerable: true, get: function () { return runtime_1.createInteropElement; } });
Object.defineProperty(exports, "StyleSheet", { enumerable: true, get: function () { return runtime_1.StyleSheet; } });
Object.defineProperty(exports, "colorScheme", { enumerable: true, get: function () { return runtime_1.colorScheme; } });
Object.defineProperty(exports, "useColorScheme", { enumerable: true, get: function () { return runtime_1.useColorScheme; } });
Object.defineProperty(exports, "useUnstableNativeVariable", { enumerable: true, get: function () { return runtime_1.useUnstableNativeVariable; } });
Object.defineProperty(exports, "vars", { enumerable: true, get: function () { return runtime_1.vars; } });
Object.defineProperty(exports, "rem", { enumerable: true, get: function () { return runtime_1.rem; } });
