// react-native-css-interop/dist/runtime/native/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rem = exports.colorScheme = exports.StyleSheet = exports.vars = exports.useUnstableNativeVariable = exports.useColorScheme = exports.remapProps = exports.cssInterop = void 0;
var api_1 = require("../web/api");
Object.defineProperty(exports, "cssInterop", { enumerable: true, get: function () { return api_1.cssInterop; } });
Object.defineProperty(exports, "remapProps", { enumerable: true, get: function () { return api_1.remapProps; } });
Object.defineProperty(exports, "useColorScheme", { enumerable: true, get: function () { return api_1.useColorScheme; } });
Object.defineProperty(exports, "useUnstableNativeVariable", { enumerable: true, get: function () { return api_1.useUnstableNativeVariable; } });
Object.defineProperty(exports, "vars", { enumerable: true, get: function () { return api_1.vars; } });
var stylesheet_1 = require("./stylesheet");
Object.defineProperty(exports, "StyleSheet", { enumerable: true, get: function () { return stylesheet_1.StyleSheet; } });
var globals_1 = require("./globals");
Object.defineProperty(exports, "colorScheme", { enumerable: true, get: function () { return globals_1.colorScheme; } });
Object.defineProperty(exports, "rem", { enumerable: true, get: function () { return globals_1.rem; } });
