// react-native-css-interop/dist/metro/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withCssInterop = exports.sendUpdate = exports.middleware = void 0;
const expo_1 = require("./expo");
var poll_updates_server_1 = require("./poll-updates-server");
Object.defineProperty(exports, "middleware", { enumerable: true, get: function () { return poll_updates_server_1.middleware; } });
Object.defineProperty(exports, "sendUpdate", { enumerable: true, get: function () { return poll_updates_server_1.sendUpdate; } });
function withCssInterop(config, options) {
    const getTransformOptions = async (...args) => {
        (0, expo_1.expoColorSchemeWarning)();
        return config.transformer?.getTransformOptions?.(...args);
    };
    return {
        ...config,
        resolver: {
            ...config.resolver,
            sourceExts: [...config?.resolver.sourceExts, "css"],
        },
        transformerPath: require.resolve("./transformer"),
        transformer: {
            ...config.transformer,
            cssToReactNativeRuntime: options,
            getTransformOptions,
            transformerPath: config.transformerPath,
        },
    };
}
exports.withCssInterop = withCssInterop;
