// react-native-css-interop/dist/src/runtime/native/globals.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isReduceMotionEnabled = exports.vh = exports.vw = exports.rem = exports.cssVariableObservable = exports.colorScheme = exports.systemColorScheme = exports.containerContext = exports.variableContext = exports.universalVariables = exports.rootVariables = exports.externalCallbackRef = exports.warned = exports.warnings = void 0;
const react_native_1 = require("react-native");
const shared_1 = require("../../shared");
const observable_1 = require("../observable");
const react_1 = require("react");
exports.warnings = new Map();
exports.warned = new Set();
exports.externalCallbackRef = {};
exports.rootVariables = {};
exports.universalVariables = {};
exports.variableContext = (0, react_1.createContext)(exports.rootVariables);
exports.containerContext = (0, react_1.createContext)({});
exports.systemColorScheme = (0, observable_1.observable)(react_native_1.Appearance.getColorScheme() ?? "light");
const colorSchemeObservable = (0, observable_1.observable)(undefined, { fallback: exports.systemColorScheme });
exports.colorScheme = {
    set(value) {
        if (value === "system") {
            colorSchemeObservable.set(undefined);
            appearance.setColorScheme(null);
        }
        else {
            colorSchemeObservable.set(value);
            appearance.setColorScheme(value);
        }
    },
    get: colorSchemeObservable.get,
    toggle() {
        let current = colorSchemeObservable.get();
        if (current === undefined)
            current = appearance.getColorScheme() ?? "light";
        colorSchemeObservable.set(current === "light" ? "dark" : "light");
    },
    [shared_1.INTERNAL_RESET]: (appearance) => {
        colorSchemeObservable.set(undefined);
        resetAppearanceListeners(appearance, react_native_1.AppState);
    },
};
function cssVariableObservable(value, { name } = {}) {
    const light = (0, observable_1.observable)(value?.light, { name: `${name}#light` });
    const dark = (0, observable_1.observable)(value?.dark, {
        name: `${name}#dark`,
        fallback: light,
    });
    return {
        get(effect) {
            return exports.colorScheme.get(effect) === "light"
                ? light.get(effect)
                : dark.get(effect);
        },
        set(value, notify = false) {
            if (typeof value === "object" && value) {
                if ("dark" in value)
                    dark.set(value.dark, notify);
                if ("light" in value)
                    light.set(value.light, notify);
            }
            else {
                light.set(value, notify);
                dark.set(value, notify);
            }
        },
    };
}
exports.cssVariableObservable = cssVariableObservable;
let appearance = react_native_1.Appearance;
let appearanceListener;
let appStateListener;
function resetAppearanceListeners($appearance, appState) {
    appearance = $appearance;
    appearanceListener?.remove();
    appStateListener?.remove();
    appearanceListener = appearance.addChangeListener((state) => {
        if (react_native_1.AppState.currentState === "active") {
            exports.systemColorScheme.set(state.colorScheme ?? "light", true);
        }
    });
    appStateListener = appState.addEventListener("change", (type) => {
        if (type === "active") {
            exports.systemColorScheme.set(appearance.getColorScheme() ?? "light", true);
        }
    });
}
resetAppearanceListeners(appearance, react_native_1.AppState);
exports.rem = (0, observable_1.observable)(14);
const viewport = (0, observable_1.observable)(react_native_1.Dimensions.get("window"));
let windowEventSubscription;
const viewportReset = (dimensions) => {
    viewport.set(dimensions.get("window"));
    windowEventSubscription?.remove();
    windowEventSubscription = dimensions.addEventListener("change", (size) => {
        return viewport.set(size.window, true);
    });
};
viewportReset(react_native_1.Dimensions);
exports.vw = {
    get: (effect) => viewport.get(effect).width,
    [shared_1.INTERNAL_RESET]: viewportReset,
    [shared_1.INTERNAL_SET](value) {
        viewport.set({ ...viewport.get(), width: value });
    },
};
exports.vh = {
    get: (effect) => viewport.get(effect).height,
    [shared_1.INTERNAL_RESET]: viewportReset,
    [shared_1.INTERNAL_SET](value) {
        viewport.set({ ...viewport.get(), height: value });
    },
};
exports.isReduceMotionEnabled = Object.assign((0, observable_1.observable)(false, { name: "isReduceMotionEnabled" }), { [shared_1.INTERNAL_RESET]: () => exports.isReduceMotionEnabled.set(false) });
react_native_1.AccessibilityInfo.isReduceMotionEnabled()?.then(exports.isReduceMotionEnabled.set);
react_native_1.AccessibilityInfo.addEventListener("reduceMotionChanged", (value) => {
    exports.isReduceMotionEnabled.set(value, true);
});
